### PR TITLE
Improve loop performance

### LINKED ISSUE #836: Improve performance of CompletionStages.loop
* Remove the `AtomicInteger`s
* Avoid use of InStream
* Avoid casting arrays to streams
* Don't chain unnecessary `CompletionStage`s

### PR DESCRIPTION
**Pull request by DavideD**
Fixes #836 

I still need to add a couple of test cases but otherwise the PR is ready for a review:

* It removes the AtomicInteger
* Avoid chaining emtpy CompletionStages when possible
* Avoid InStream
* Avoid casting arrays to Stream
* Avoid iterators where possible

### CODE REVIEW

**Comment by DavideD** [comment]
I've added the missing tests and this is now ready to be reviewed

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/Cascade.java [hunk 1]
```diff
@@ -483,7 +484,7 @@ private void cascadeToOne(
 			persistenceContext.addChildParent( child, parent );
 			stage = stage.thenCompose( v -> action.cascade( eventSource, child, entityName, context, isCascadeDeleteEnabled ) )
 					.whenComplete( (vv, e) -> persistenceContext.removeChildParent( child ) )
-					.thenAccept( vv -> {} );
+					.thenCompose( CompletionStages::voidFuture );
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/Cascade.java:487]: I just sent a PR that makes this change obsolete.

**Review by gavinking — APPROVED** [review]
LGTM!

**Comment by gavinking** [comment]
But please don't forget to squash ;-)

**Comment by DavideD** [comment]
> But please don't forget to squash ;-)

Sure, on Monday I will reduce the commits. Have a nice week-end

**Comment by gavinking** [comment]
You too :)

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java [hunk 2]
```diff
@@ -337,59 +337,56 @@ private void disallowDeletionOfDetached(DeleteEvent event) {
...
+										entityEntry.getDeletedState(),
 										persister,
-										isCascadeDeleteEnabled,
-										session
-								)
-						);
-					}
-				} );
-
-		CompletionStage<Void> afterDelete = cascadeAfterDelete( session, persister, entity, transientEntities );
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java:390]: Hrm. @Davide I actually had this being called outside of the `thenCompose()` callback intentionally.

So this change actually changes the behavior of the code somewhat.

It probably doesn't break anything but I'm still not sure I like it.

**Comment by DavideD** [comment]
Ok, but you cannot rely on the order of the operations if we don't chain
them this way ( theoretically). Why would you want to keep them separated?

**Comment by gavinking** [comment]
> Ok, but you cannot rely on the order of the operations if we don't chain
them this way ( theoretically).

Well I think the ordering is still well-defined. All the non-async processing happens up-front, before any async process, and then the async stuff happens, with a well-defined order.

>  Why would you want to keep them separated?

Well, for one thing, imagine how impossible it's going to be to debug the cascading logic if it's happening over a bunch of async callbacks. I did it this way because I liked the idea of just doing all the sync stuff upfront.

**Comment by DavideD** [comment]
I didn't expect this to be a controversial commit. I will remove it from this PR because it's not really related and I don't follow  you right now.

I'm going to have a better look at it in the future.

**Comment by gavinking** [comment]
Yeah fine. I'm not completely against the change, I just think it bears more thinking-through.

**Comment by DavideD** [comment]
Squashed things together, waiting for CI

**Pull request merged**